File: code/05-dynamic-programming/python/max_subarray.py
"""Same recurrence as the "Brute Force" pseudocode and the bestEndingAt/
bestOverall DP template (Kadane), and the C/Java demos in this chapter --
newly written in Python, no canonical source exists."""


# snippet:max-subarray-brute-force:start
def max_subarray_brute_force(a):
    best_sum, best_start, best_end = a[0], 0, 0
    for i in range(len(a)):
        running = 0
        for j in range(i, len(a)):
            running += a[j]
            if running > best_sum:
                best_sum, best_start, best_end = running, i, j
    return best_sum, best_start, best_end
# snippet:max-subarray-brute-force:end


def _length(start, end):
    return end - start + 1


def _better(a, b):
    a_sum, a_start, a_end = a
    b_sum, b_start, b_end = b
    if a_sum != b_sum:
        return a_sum > b_sum
    a_len, b_len = _length(a_start, a_end), _length(b_start, b_end)
    if a_len != b_len:
        return a_len < b_len
    return a_start < b_start


# snippet:max-subarray-kadane:start
def max_subarray_kadane(a):
    ending = (a[0], 0, 0)
    best = ending
    for i in range(1, len(a)):
        extend = (ending[0] + a[i], ending[1], i)
        restart = (a[i], i, i)
        ending = restart if _better(restart, extend) else extend
        if _better(ending, best):
            best = ending
    return best
# snippet:max-subarray-kadane:end


if __name__ == "__main__":
    a = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    bf_sum, bf_start, bf_end = max_subarray_brute_force(a)
    kd_sum, kd_start, kd_end = max_subarray_kadane(a)
    print("brute_force sum:", bf_sum, "start:", bf_start, "end:", bf_end)
    print("kadane sum:", kd_sum, "start:", kd_start, "end:", kd_end)

File: code/05-dynamic-programming/c/max_subarray.c
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>

typedef struct { long long sum; size_t start, end; } MaxSubarrayResult;

/* All Theta(n^2) intervals, matching the "Brute Force" pseudocode exactly
 * (every (i,j) pair, running sum). New for this book -- the canonical file
 * only has the linear-time Kadane form. */

// snippet:max-subarray-brute-force:start
static bool max_subarray_brute_force(const long long *a, size_t n, MaxSubarrayResult *out) {
    if (!a || !out || n == 0) return false;
    MaxSubarrayResult best = {a[0], 0, 0};
    for (size_t i = 0; i < n; ++i) {
        long long sum = 0;
        for (size_t j = i; j < n; ++j) {
            sum += a[j];
            if (sum > best.sum) { best.sum = sum; best.start = i; best.end = j; }
        }
    }
    *out = best;
    return true;
}
// snippet:max-subarray-brute-force:end

static size_t length(MaxSubarrayResult r){return r.end-r.start+1;}
static bool better(MaxSubarrayResult a, MaxSubarrayResult b){
    return a.sum>b.sum||(a.sum==b.sum&&(length(a)<length(b)||(length(a)==length(b)&&a.start<b.start)));
}

/* bestEndingAt[i]/bestOverall (Kadane), matching the DP-template state split.
 * Reused verbatim from lecture-notes/code/lecture05/c/max_subarray.c
 * (max_subarray), renamed to match this chapter's section heading. Nonempty,
 * 0-based inclusive interval. Theta(n) time, Theta(1) space. */

// snippet:max-subarray-kadane:start
static bool max_subarray_kadane(const long long *a,size_t n,MaxSubarrayResult *out){
    if(!a||!out||n==0)return false;
    MaxSubarrayResult ending={a[0],0,0},best=ending;
    for(size_t i=1;i<n;++i){
        MaxSubarrayResult extend={ending.sum+a[i],ending.start,i},restart={a[i],i,i};
        ending=better(restart,extend)?restart:extend;if(better(ending,best))best=ending;
    }*out=best;return true;
}
// snippet:max-subarray-kadane:end

int main(void) {
    const long long a[] = {-2,1,-3,4,-1,2,1,-5,4};
    size_t n = 9;
    MaxSubarrayResult r1, r2;
    max_subarray_brute_force(a, n, &r1);
    max_subarray_kadane(a, n, &r2);
    printf("brute_force sum: %lld start: %zu end: %zu\n", r1.sum, r1.start, r1.end);
    printf("kadane sum: %lld start: %zu end: %zu\n", r2.sum, r2.start, r2.end);
    return 0;
}
